### PR TITLE
Specify `log_root` in `imitation.scripts.common` to replace default `log_root`

### PR DESCRIPTION
**Pull request by yawen-d**
## Description

Address #526.

## Testing

Existing tests already suffice.

### CODE REVIEW

File: src/imitation/scripts/common/common.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def hook(config, command_name, logger):
     if config["common"]["log_dir"] is None:
         env_sanitized = config["common"]["env_name"].replace("/", "_")
         log_dir = os.path.join(
-            "output",
+            config["common"]["log_root"] or "output",
```
Comment (AdamGleave) [inline src/imitation/scripts/common/common.py:52]: Should maybe define this as a separate variable since the logic is a little complex?

Comment (yawen-d) [reply src/imitation/scripts/common/common.py:52]: yep

**Comment by AdamGleave** [comment]
Tests are now failing on Windows because `TemporaryDirectory` used in https://github.com/HumanCompatibleAI/imitation/blob/yawen-d/fix/add-log_root/tests/scripts/test_scripts.py#L830 tries to remove the directory after the script returns, but there's still an open file handle to Monitor. This isn't a problem on Unix that allows files to be unlinked that are still open by a process, but is a problem on Windows. The file handle is still open because we never actually call `venv.close()`. I've modified the scripts to close this using a `try:`, `finally:` block similar to what we were already doing in `eval_policy`.

Worried what I've done is a bit overkill, it is annoying that most of the logic is now in a `try:` indent block, we could also just have `venv.close()` outside `try-finally` and accept it won't get cleaned up on exceptions.

**Comment by Rocamonde** [comment]
> Worried what I've done is a bit overkill, it is annoying that most of the logic is now in a `try:` indent block, we could also just have `venv.close()` outside `try-finally` and accept it won't get cleaned up on exceptions.

Another option is to create a context manager that automatically closes the environment and then enter it using a `with` statement. If this is done in many places across the repo, it could be a good use case for abstracting it out.

File: src/imitation/scripts/train_rl.py [hunk 2]
```diff
@@ -96,55 +96,66 @@ def train_rl(
...
+                rollout_save_n_timesteps,
+                rollout_save_n_episodes,
+            )
+            types.save(save_path, rollout.rollout(rl_algo, venv, sample_until))
+        if policy_save_final:
+            output_dir = os.path.join(policy_dir, "final")
+            serialize.save_stable_model(output_dir, rl_algo)
+
+        # Final evaluation of expert policy.
+        result = train.eval_policy(rl_algo, venv)
```
Comment (Rocamonde) [inline src/imitation/scripts/train_rl.py:154]: This is not necessary, you can return, and the finally still gets called anyways. (e.g. https://stackoverflow.com/questions/19805654/python-try-finally-block-returns)

**Review by Rocamonde — COMMENTED** [review]
LGTM, apart from the minor style suggestion.

**Review by Rocamonde — APPROVED** [review]

**Comment by AdamGleave** [comment]
Good suggestion to use context manager instead. I've changed `common.make_venv` to be a context manager and updated the scripts to use `with` instead of `try:`, `finally:`.

I debated whether to turn `util.make_vec_env` (which `common.make_venv` is a thin wrapper around) into a context manager but ultimately decided against. Mandating use of context manager seemed like it constrains end-user code more than we really want. An alternative would be to have `util.make_vec_env` return an object that can be used directly or via a context-manager, ala `open`, but I think this would require some hacky wrapping of `VecEnv` objects to add `__enter__` and `__exit__` methods -- probably not worth it.

**Review by Rocamonde — APPROVED** [review]
LGTM.

File: src/imitation/scripts/train_imitation.py [hunk 3]
```diff
@@ -119,52 +119,55 @@ def train_imitation(
...
+            )
+            # TODO(adam): add checkpointing to DAgger?
+            save_locations = model.save_trainer()
+            print(f"Model saved to {save_locations}")
+        else:
+            bc_trainer.train(**bc_train_kwargs)
+            # TODO(adam): add checkpointing to BC?
+            bc_trainer.save_policy(policy_path=osp.join(log_dir, "final.th"))
+
+        imit_stats = train.eval_policy(imit_policy, venv)
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/scripts/train_imitation.py:167]: If I was just doing `return imit_stats` in the next line I'd agree. But we also have code computing stats from expert rollouts that's outside the context manager (but part of return value). I left it outside to minimize the surface of the code inside the context manager. This doesn't particularly matter, I can see an argument for the other way, but it seems nice to draw a distinction between code that does vs doesn't need the venv as a form of self-documentation.

**Pull request merged**